magic_marker/magic_marker.py

```diff
@@ -0,0 +1,199 @@
+
+    def _backup_whole_dir(self, directory):
+        """Backup the entire target
+
+        Args:
+            directory (str): the path to copy
+        """
+        try:
+            shutil.copytree(directory, self._backup_dir)
+        except OSError as exc:  # python >2.5
+            if exc.errno == errno.ENOTDIR:
+                # its a single file
+                shutil.copy(directory, self.backup_dir)
+            else:
+                raise RuntimeError("Magic Marker was not able to backup the target {}".format(directory))
+
+    def fix_it(self, flake8_output):
+        """Perform the fix
+
+        Args:
+            flake8_output (str): The parsed json data from flake8-json
+
+        Returns:
+            str: the message stating what was performed
+        """
+        message = ""
+        for file_path in flake8_output:
+            if flake8_output[file_path]:
+                filname, fixcount = self._fix_file(flake8_output[file_path])
+                if fixcount == 1:
+                    message += "\n{} : {} test marked with UUID".format(filname, fixcount)
+                else:
+                    message += "\n{} : {} tests marked with UUID".format(filname, fixcount)
+        if message:
+            return message
+        else:
+            return "No changes made"
+
+    def _fix_file(self, fixes_required):
+        """Fixes an individual file
+
+        Args:
+            fixes_required (list[dict]): a list of the fixes required for this file
+
+        Returns:
+            tuple(str, str):  filename and number of fixes performed
+        """
+        filename = str(fixes_required[0]['filename'])
+        with open(filename, 'r') as f:
+            data = f.readlines()
+
+        fixes_required.sort(key=lambda x: x['line_number'])
+        fixes_required.reverse()  # work from the bottom of the file up
+        for fix in fixes_required:
+            fix_position = fix['line_number']
+            if fix_position:
+                fix_position += -1
+            mark = self._match_indent(data[fix_position], self._uuid_mark())
+            data.insert(fix_position, mark)
+
+        with open(filename, 'w') as f:
+            f.writelines(data)
+        return filename, len(fixes_required)
+
+    def _uuid_mark(self):
+        """generate a UUID mark string
+
+        Returns:
+            str: the correctly formatted UUID mark string
+        """
+        return "@pytest.mark.{}('{}')\n".format(self._mark, str(uuid.uuid1()))
+
+    @staticmethod
+    def _match_indent(def_string, target_line):
+        """Match the indent of the original string
+
+        Args:
+            def_string (str): the original string from the test definition
+            target_line (str): the string that should be updated
+
+        Returns:
+            str : The new string updated with correct whitespace
+        """
+        regex = re.compile('(^\s*)')
+        whitespace = re.match(regex, def_string).group(1)
```

Comment:
Regex compile unnecessary given that it will be recompiled every time this method gets called.